source/azure_iot.c

```diff
@@ -0,0 +1,438 @@
+
+AzureIoTResult_t AzureIoT_MessagePropertiesFind( AzureIoTMessageProperties_t * pxMessageProperties,
+                                                 const uint8_t * pucName,
+                                                 uint32_t ulNameLength,
+                                                 const uint8_t ** ppucOutValue,
+                                                 uint32_t * pulOutValueLength )
+{
+    az_span xNameSpan = az_span_create( ( uint8_t * ) pucName, ( int32_t ) ulNameLength );
+    az_span xOutValueSpan;
+    az_result xResult;
+
+    if( ( pxMessageProperties == NULL ) ||
+        ( ppucOutValue == NULL ) || ( pulOutValueLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoTMessagePropertiesFind failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xResult = az_iot_message_properties_find( &pxMessageProperties->_internal.xProperties,
+                                              xNameSpan, &xOutValueSpan );
+
+    if( az_result_failed( xResult ) )
+    {
+        return eAzureIoTItemNotFound;
+    }
+
+    *ppucOutValue = az_span_ptr( xOutValueSpan );
+    *pulOutValueLength = ( uint32_t ) az_span_size( xOutValueSpan );
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Base64HMACCalculate( AzureIoTGetHMACFunc_t xAzureIoTHMACFunction,
+                                               const uint8_t * pucKey,
+                                               uint32_t ulKeySize,
+                                               const uint8_t * pucMessage,
+                                               uint32_t ulMessageSize,
+                                               uint8_t * pucBuffer,
+                                               uint32_t ulBufferLength,
+                                               uint8_t * pucOutput,
+                                               uint32_t ulOutputSize,
+                                               uint32_t * pulOutputLength )
+{
+    AzureIoTResult_t xStatus;
+    uint8_t * pucHashBuf;
+    uint32_t ulHashBufSize = AZURE_BASE64_HASH_BUFFER_SIZE;
+    uint32_t ulBinaryKeyBufSize;
+
+    if( ( xAzureIoTHMACFunction == NULL ) ||
+        ( pucKey == NULL ) || ( ulKeySize == 0 ) ||
+        ( pucMessage == NULL ) || ( ulMessageSize == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( pucOutput == NULL ) || ( pulOutputLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoT_Base64HMACCalculate failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    ulBinaryKeyBufSize = ulBufferLength;
+    xStatus = prvAzureIoTBase64Decode( ( char * ) pucKey, ulKeySize,
+                                       pucBuffer, ulBinaryKeyBufSize, &ulBinaryKeyBufSize );
+
+    if( xStatus )
+    {
+        return xStatus;
+    }
+
+    ulBufferLength -= ulBinaryKeyBufSize;
```

Comment:
Nit: Whenever there's subtraction I always worry we'll end up with underflow, so ulBufferLength would be negative (or giant unsigned int) if ulBinaryKeyBufSize is bigger.

That's not case here because the decoded length will be smaller, but it might not be bad to assert if we do such things about this or put a comment to this effect?

Reply:
added a comment. let me know if it's sufficient

Reply:
yup, resolving.